File: Baekjoon/Class3/21.1654 랜선 자르기.java
import java.io.*;
import java.util.StringTokenizer;

class Main {
    private static long countCutWire(long length, long[] wires) { // 잘린 선 개수
        long count = 0;
        for(long wire: wires)
            count += wire / length;
        return count;
    }

    private static long maxCutLength(int required, long[] wires) {
        long answer = 0;
        long left = 1L, right = wires[0];
        for(int i = 1; i < wires.length; i++)
            right = Math.max(right, wires[i]);
        while(left <= right) {
            long middle = left + (right - left) / 2L;
            if(countCutWire(middle, wires) >= required) { // 더 길게 자를 수도 있음
                left = middle + 1L;
                answer = middle;
            }
            else // 더 짧게 잘라야 함
                right = middle - 1L;
        }
        return answer;
    }

    public static void main(String[] args) throws IOException {
        int k, n;
        long[] wires;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        k = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        wires = new long[k];
        for(int i = 0; i < k; i++)
            wires[i] = Integer.parseInt(br.readLine());

        System.out.print(maxCutLength(n, wires));
        br.close();
    }
}

File: Baekjoon/Class3/21.1654 랜선 자르기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

long long count_cut_wire(long long length, const vector<long long>& wires) { // 잘린 선 개수
    long long count = 0;
    for (long long wire : wires)
        count += wire / length;
    return count;
}

long long max_cut_length(int required, const vector<long long>& wires) {
    long long answer = 0;
    long long left = 1LL, right = *max_element(wires.begin(), wires.end());
    while (left <= right) {
        long long middle = left + (right - left) / 2LL;
        if (count_cut_wire(middle, wires) >= required) { // 더 길게 자를 수도 있음
            left = middle + 1LL;
            answer = middle;
        }
        else // 더 짧게 잘라야 함
            right = middle - 1LL;
    }
    return answer;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int k, n;
    cin >> k >> n;

    vector<long long> wires(k);
    for (int i = 0; i < k; i++)
        cin >> wires[i];

    cout << max_cut_length(n, wires);
    return 0;
}